Application Screenshots included

traliman opened the pull request:
New property IncludeScreenshots in ISettings is made. New class ScreenShotWriter is made, ReportStorage is  modified to include screenshots in zip file. Examples is  modified to see the local copies of zip files(LocalSender class).

traliman committed:
Screenshots included

akurdyukov commented on Examples/ConsoleExample/LocalSender.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace NCrash.Examples.ConsoleExample
+{
+    class localsender : ISender
```
Naming rules violation. Class name should start with capital letter. This class should be 'LocalSender'.

akurdyukov commented on Examples/ConsoleExample/LocalSender.cs:
```diff
@@ -0,0 +1,30 @@
+        public bool Send(Stream data, string fileName, Report report)
+        {
+
+            filew = new FileStream(fileName, FileMode.Create);
```
This code will cause problems if something between FileStream creation and Close throws exception. Need 'using' here.

akurdyukov commented on Examples/ConsoleExample/Program.cs:
```diff
@@ -10,7 +10,13 @@ static void Main()
             var reporter = new ErrorReporter(settings);
+            reporter.HandleExceptions = true;
+            //System.IO.IsolatedStorage.IsolatedStorageFile.Remove(System.IO.IsolatedStorage.IsolatedStorageScope.User);
+            //mycode
```
Cannot merge debug comments.

akurdyukov commented on Examples/WinFormsExample/LocalSender.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace NCrash.Examples.WinFormsExample
+{
+    class localsender : ISender
```
Same comments as above.

akurdyukov commented on NCrash/Core/ScreenShots/ScreenShotWriter.cs:
```diff
@@ -0,0 +1,98 @@
+﻿//mycode
```
This comment is not useful

akurdyukov commented on NCrash/Core/ScreenShots/ScreenShotWriter.cs:
```diff
@@ -0,0 +1,98 @@
+            public int Bottom;
+        }
+
+        private static List<int> Windows = new List<int>();
```
invalid naming, should me '_windows'. See other classes for style.
Also, too wide type, you can create IList here

akurdyukov commented on NCrash/Core/ScreenShots/ScreenShotWriter.cs:
```diff
@@ -0,0 +1,98 @@
+                {
+
+                    screenshot = Capture(hwnd);
+                    screenshotName = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData),
```
This path should be tunable

akurdyukov commented on NCrash/Core/ScreenShots/ScreenShotWriter.cs:
```diff
@@ -0,0 +1,98 @@
+                    screenshotNames.Add(screenshotName);
+                }
+            }
+            catch (Exception e)
```
Very bad idea to catch exception, but show nothing.

akurdyukov commented on NCrash/Storage/ReportStorage.cs:
```diff
@@ -54,6 +55,22 @@ public void Write(Report report)
+                        if (_settings.IncludeScreenshots)
+                        {
+                            int i = 1;
+                            System.Collections.Generic.List<string> names =  ScreenShotWriter.Write();
```
Why use full class reference?

traliman committed:
Screenshots included update

traliman committed:
include screenshots update

traliman committed:
include screenshots update

traliman committed:
Comments add

traliman commented:
Added unserialized property -  the list of screenshot files to report. Screenshots are generated in ErrorReporter.Report(). The user can attach them to zip file in  ProcessingException event handler. Added the example how to do this.

akurdyukov commented on Examples/ConsoleExample/Program.cs:
```diff
@@ -10,7 +12,19 @@ static void Main()
+            {
+                if (settings.AdditionalReportFiles == null)
+                    settings.AdditionalReportFiles = new List<string>();
+                foreach (Tuple<string, string> screenshot in report.ScreenshotList)
```
why this adding is here, not inside ErrorReporter class?

akurdyukov commented on NCrash/Core/ScreenShots/ScreenShotWriter.cs:
```diff
@@ -0,0 +1,130 @@
+        ///Geneating screenshots
+        /// </summary>
+
+        internal static IList<Tuple<string,string>> Write(string _path)
```
Underscore (_) is for private fields, not arguments.

akurdyukov commented on NCrash/ErrorReporter.cs:
```diff
@@ -267,6 +268,8 @@ protected virtual ExecutionFlow Report(Exception exception)
                 var serializableException = new SerializableException(exception);
                 var report = new Report(serializableException);
+                //Generating screenshots, adding list of result files to report.
+                report.ScreenshotList = ScreenShotWriter.Write(string.Empty);
```
Why create special fields and not use standard AdditionalFiles? You still have screenshot-specific code in generic class.

akurdyukov commented on NCrash/Core/Report.cs:
```diff
@@ -39,5 +41,11 @@ public override string ToString()
+        /// Unserialized property. List of screenshot files. Path,file.
+        /// </summary>
+        [XmlIgnoreAttribute]
+        public IList<Tuple<string, string>> ScreenshotList{get;set;}
```
See code formatting in "{get;set;}". Does not match other code style.

traliman committed:
Plugins added. ScreenShotWriter plugin released.

traliman committed:
update

traliman commented:
Plugins added. Processing and postprocessing plugins in ErrorReporter.Report. ScreenShotWriter plugin released.

akurdyukov commented on NCrash/ISettings.cs:
```diff
@@ -37,7 +38,7 @@ public interface ISettings
         /// Paths should be absolute i.e. start with disk name.
         /// </summary>
-        IList<string> AdditionalReportFiles { get; }
+        IList<string> AdditionalReportFiles { get; set; }
```
Why user setter here?

traliman replied:
If we try to assign this property in any plugin class, we get the compiller error, example:
public void PreProcess(ISettings settings)
        {
....
                if(settings.AdditionalReportFiles == null)
                    settings.AdditionalReportFiles = new List<string>();

Error   1   Property or indexer 'NCrash.ISettings.AdditionalReportFiles' cannot be assigned to -- it is read only

akurdyukov commented on NCrash/Plugins/IPlugin.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace NCrash.Plugins
+{
+    public interface IPlugin
```
Please add some docs over here

akurdyukov commented on NCrash/Plugins/ScreenShotWriter.cs:
```diff
@@ -0,0 +1,142 @@
+        /// Generating windows handle list
+        /// </summary>
+
+        internal bool AddWnd(int hwnd, int lparam)
```
This may be private

akurdyukov commented on NCrash/Plugins/ScreenShotWriter.cs:
```diff
@@ -0,0 +1,142 @@
+        /// Capturing window area at the screen
+        /// </summary>
+
+        internal Bitmap Capture(int hwnd)
```
This may be private

akurdyukov commented on NCrash/Plugins/ScreenShotWriter.cs:
```diff
@@ -0,0 +1,142 @@
+    public class ScreenShotWriter:IPlugin
+    {
+
+        public ScreenShotWriter()
```
You'd better user standard class code structure like
1. consts
2. static
3. readonly fields
4. other fields
5. constructors
6. events
7. properties
8. methods

traliman committed:
update

traliman committed:
update
